compiler/src/dotty/tools/dotc/transform/PCPCheckAndHeal.scala

```diff
@@ -0,0 +1,219 @@
+    tree.cast(tp).withSpan(tree.span)
+  }
+
+  /** If `tree` refers to a locally defined symbol (either directly, or in a pickled type),
+   *  check that its staging level matches the current level. References to types
+   *  that are phase-incorrect can still be healed as follows:
+   *
+   *  If `T` is a reference to a type at the wrong level, try to heal it by replacing it with
+   *  `~implicitly[quoted.Type[T]]`.
+   */
+  protected def checkLevel(tree: Tree)(implicit ctx: Context): Tree = {
+    def checkTp(tp: Type): Type = checkType(tree.sourcePos).apply(tp)
+    tree match {
+      case Quoted(_) | Spliced(_)  =>
+        tree
+      case tree: RefTree if tree.symbol.is(InlineParam) =>
+        tree
+      case _: This =>
+        assert(checkSymLevel(tree.symbol, tree.tpe, tree.sourcePos).isEmpty)
+        tree
+      case _: Ident =>
+        checkSymLevel(tree.symbol, tree.tpe, tree.sourcePos) match {
+          case Some(tpRef) => tpRef
+          case _ => tree
+        }
+      case _: TypeTree | _: AppliedTypeTree | _: Apply | _: TypeApply | _: UnApply | Select(_, OuterSelectName(_, _)) =>
+        tree.withType(checkTp(tree.tpe))
+      case _: ValOrDefDef | _: Bind =>
+        tree.symbol.info = checkTp(tree.symbol.info)
+        tree
+      case _: Template =>
+        checkTp(tree.symbol.owner.asClass.givenSelfType)
+        tree
+      case _ =>
+        tree
+    }
+  }
+
+  /** Check and heal all named types and this-types in a given type for phase consistency. */
+  private def checkType(pos: SourcePosition)(implicit ctx: Context): TypeMap = new TypeMap {
+    def apply(tp: Type): Type = reporting.trace(i"check type level $tp at $level") {
+      tp match {
+        case tp: TypeRef if tp.symbol.isSplice =>
+          if (tp.isTerm)
+            ctx.error(i"splice outside quotes", pos)
+          tp
+        case tp: NamedType =>
+          checkSymLevel(tp.symbol, tp, pos) match {
+            case Some(tpRef) => tpRef.tpe
+            case _ =>
+              if (tp.symbol.is(Param)) tp
+              else mapOver(tp)
+          }
+        case tp: ThisType =>
+          assert(checkSymLevel(tp.cls, tp, pos).isEmpty)
+          mapOver(tp)
+        case _ =>
+          mapOver(tp)
+      }
+    }
+  }
+
+  /** Check reference to `sym` for phase consistency, where `tp` is the underlying type
+   *  by which we refer to `sym`. If it is an inconsistent type try construct a healed type for it.
+   *
+   *  @return `None` if the phase is correct or cannot be healed
+   *          `Some(tree)` with the `tree` of the healed type tree for `~implicitly[quoted.Type[T]]`
+   */
+  private def checkSymLevel(sym: Symbol, tp: Type, pos: SourcePosition)(implicit ctx: Context): Option[Tree] = {
+    val isThis = tp.isInstanceOf[ThisType]
+    if (!isThis && !sym.is(Param) && sym.maybeOwner.isType)
+      None
+    else if (sym.exists && !sym.isStaticOwner && !levelOK(sym))
+      tryHeal(sym, tp, pos)
```

Comment:
The logic here is very hard to follow, to the point where I cannot be sure it is correct. 
A clearer formulation would be:
```
val needsCheck =
  tp.isInstanceOf[ThisType] || sym.is(Param) || sym.owner.isTerm
if (sym.exists && needsCheck && !sym.isStaticOwner && !levelOK(sym)) 
  tryHeal(sym, tp, pos)
else
  None
```
and it would be good to have further explanations in comments.

Reply:
I partially refactored it.

Reply:
`shouldtryHealable` can be a `def`, since it is only called after 3 other conditions. 
It would be good to add a short comment to each of the 3 lines of shouldTryHealable, indicating why we
should try to heal in this case.

Reply:
After playing around with this condition I figured out that is was checking is the symbol is a reference to a class (but not a `this.type`). I updated the code accordingly.

Reply:
Ah, that clarifies things!